cmd/spire-server/cli/entry/show.go

```diff
@@ -134,6 +133,13 @@ func (c *showCommand) fetchEntries(ctx context.Context, client entry.EntryClient
 		}
 	}
 
+	if len(c.federatesWith) != 0 {
+		filter.ByFederatesWith = &types.FederatesWithMatch{
+			TrustDomains: c.federatesWith,
+			Match:        types.FederatesWithMatch_MATCH_EXACT,
```

Comment:
I think this is a change in matching behavior? Was it intentional?

I am not sure what the desired behavior is (I guess I'd have to think about that a little more), but thought I'd comment on it.

Reply:
I see now that's a change in the current behavior for the `-federatesWith` flag. At the same time, when running this command filtering by selectors (`-selectos` flag), we have always done an exact match which I guess skewed my judgement when making the change here. Maybe it makes more sense to have an exact match in selectors and not in federated domains, but it feels kind of weird to have different behaviors.
If we want to keep strict compatibility with the previous behavior, I should change this to a subset match. We may also introduce an additional flag to change the default behavior if we feel it's needed.

Reply:
I think we may use subset match here, if for example we have:
```
entry1 : {'fedated1', 'federated2'}
entry2 : {'fedated1', 'federated3'}
```
 and we filter by `federated1`

It will returns not entries, and I think the intention on we adding this filter was to be able to found, all entries that has certain federation, if it is a problem we maybe can add some configuration, so people can choose if they want match or subset for this.

Reply:
So, let's call the CLI selectors `CS` and the entry selectors `ES` assuming the following definitions (which match API behavior, with exception of SUPERSET, which does not exist):

```
MATCH_EXACT:   Matches all entries where ES matches CS
MATCH_SUBSET: Matches all entries where ES is an exact subset of CS
MATCH_SUPERSET: Matches all entries where ES is an exact superset of CS
```

`MATCH_SUBSET` is equivalent to what happens with workload/node attestation today. The selectors presented for the agent or workload are `CS` and any entry with entries `ES`, where `ES` is a subset of `CS`, is matched.

Today's CLI behavior, regarding `federatesWith` is none of the above. Instead it does CLI-side filtering where basically it matches all entries where ES has any overlap with CS, i.e. where the intersection of ES and CS is not the empty set.

Reply:
Thanks for the clarification, @azdagron.
I think that ideally the CLI behavior should not depart from the API behavior. Since we are about to release 1.0.0, maybe this is the opportunity to introduce a change in the current behavior to be consistent with the API. Also, adding a `-matchBehavior` flag in this PR would be simple.
One path to solve this now could be:
1. Change the default behavior in the CLI to use `MATCH_SUBSET` in both `-federatesWith` and `-selectos`.
2. Add a `-matchBehavior` flag which would accept `MATCH_EXACT` and `MATCH_SUBSET` as possible values. The description of the behavior should be expressed clearly in the documentation / output.

We can consider to add a `MATCH_SUPERSET` type of filter in a future release.

Thoughts?

Reply:
I'm not sure I'm comfortable with a breaking change to the behavior without a deprecation path but also struggle to see how we can migrate this behavior meaningfully. I guess 1.0.0 is the time to do it if ever, but this brings up a larger discussion of CLI behavior migration strategies.

That aside, I'm generally ok with the proposal, although I'd probably nitpick on the cli flag and options. Maybe something like `-matchOn` with the values `subset` and `exact`?

Reply:
How about adding a matching mode of "ANY" which covers 

> Today's CLI behavior, regarding federatesWith is none of the above. Instead it does CLI-side filtering where basically it matches all entries where ES has any overlap with CS, i.e. where the intersection of ES and CS is not the empty set.

... and I agree with @azdagron nitpick on the cli flags: -matchOn with values any, exact, subset (and superset, when requested) feels clean

Reply:
> So, let's call the CLI selectors CS and the entry selectors ES

Thank you for these definitions, they are very useful :)

> I think that ideally the CLI behavior should not depart from the API behavior.

I feel that we never really gave the matching semantics the attention they deserved from a consumption standpoint. These things were put in a long time ago, and we were dealing with a lot of different types of matching semantics across the codebase... and I think that shows in the current behavior. 

Nominally, API behavior is a function of the needs of its consumers, where CLI is one of them. I think what this means is that our desired CLI behavior should influence the API behavior, and not the other way around.

Most of the time, it's a human using the CLI. I think that in practice, humans are unlikely to know the full selector set of something they're looking for... instead, they'll probably know one piece or another (e.g. I know the k8s service account name but not the image ID). If that's the case, then the most useful CLI match is `MATCH_SUPERSET`

@MarcosDY's example of federation I think supports this - I want to know all workloads that federate with X... maybe because I want to delete X, maybe because X was compromised, etc. I don't think these questions can be answered with `MATCH_EXACT` or `MATCH_SUBSET`.

> I'm not sure I'm comfortable with a breaking change to the behavior without a deprecation path but also struggle to see how we can migrate this behavior meaningfully.

I agree, the default CLI behaviors are hard to change. As a result, if we're going to change them, we'll want to have a high degree of confidence that the new default won't need to be changed again, at least for a "long time". What that means for this PR is that either we need to retain the default behavior somehow (e.g. `-matchOn` flag with `MATCH_ANY` default), or we'll want to have a strong set of use cases justifying the new behavior. I can think of only one good use case for `MATCH_SUBSET`, and that is "I have a set of selectors that I got from the agent logs, and I want to query for entries in the server that might match it".

Reply:
I agree, @evan2645. I think SUPERSET is the most natural behavior from a CLI perspective and that SUBSET has very limited usefulness.

I think I'd be most comfortable with maintaining current behavior for pre-1.0 and introducing a new default in 1.0. There are other breaking changes we are likely planning for that monumental release, so seems like a place where would-be adopters are likely to already be scanning the changelog for behavior changes before updating.